### PR TITLE
Introduce ValueFetchContext

### PR DESCRIPTION
**Pull request by cbuescher**
This PR starts pulling out some methods used in value fetchers into a new ValueFetchContext interface that should eventually be used to replace direct access to SearchExecutionContext for better isolation when introducing per segment concurrency. Currently wip only.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java [hunk 1]
```diff
@@ -44,13 +44,13 @@ public class FieldFetcher {
      */
     private static final int AUTOMATON_MAX_DETERMINIZED_STATES = 100000;
 
-    public static FieldFetcher create(SearchExecutionContext context, Collection<FieldAndFormat> fieldAndFormats) {
+    public static FieldFetcher create(ValueFetchContext context, Collection<FieldAndFormat> fieldAndFormats) {
         Set<String> nestedMappingPaths = context.nestedLookup().getNestedMappers().keySet();
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:48]: Is this the only place `nestedLookup()` is used?  I'd be tempted to try and pull it out into FetchContext and then pass NestedLookup as a separate parameter here?

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java [hunk 2]
```diff
@@ -105,14 +105,14 @@ public HighlightField highlight(FieldHighlightContext fieldContext) throws IOExc
         int numberOfFragments = field.fieldOptions().numberOfFragments() == 0 ? 1 : field.fieldOptions().numberOfFragments();
         ArrayList<OrderedTextFragment> fragsList = new ArrayList<>();
         List<Object> textsToHighlight;
-        final int maxAnalyzedOffset = context.getSearchExecutionContext().getIndexSettings().getHighlightMaxAnalyzedOffset();
+        final int maxAnalyzedOffset = context.getValueFetchContext().getIndexSettings().getHighlightMaxAnalyzedOffset();
         Integer queryMaxAnalyzedOffset = fieldContext.field.fieldOptions().maxAnalyzedOffset();
         Analyzer analyzer = wrapAnalyzer(
-            context.getSearchExecutionContext().getIndexAnalyzer(f -> Lucene.KEYWORD_ANALYZER),
+            context.getValueFetchContext().getIndexAnalyzer(f -> Lucene.KEYWORD_ANALYZER),
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java:111]: I'd be tempted to try and move both `getIndexSettings()` and `getIndexAnalyzer()` to `FetchContext`

**Review by romseygeek — COMMENTED** [review]
This is a great start, thanks @cbuescher.  I left some initial comments.

**Comment by cbuescher** [comment]
No work here currently, so I'm closing this.

**Pull request closed without merging**